test/file/main.c

```diff
@@ -203,6 +215,236 @@ static int __test_posix_fallocate(const char *file_path) {
     return 0;
 }
 
+#ifndef __GLIBC__
+#define FALLOC_FL_COLLAPSE_RANGE (0x08)
+#define FALLOC_FL_ZERO_RANGE (0x10)
+#define FALLOC_FL_INSERT_RANGE (0x20)
+#define FALLOC_FL_UNSHARE_RANGE (0x40)
+#endif
+
+static int __test_fallocate_with_invalid_mode(const char *file_path) {
+    int fd = open(file_path, O_RDWR);
+    if (fd < 0) {
+        THROW_ERROR("failed to open a file to read/write");
+    }
+    off_t len = 2 * BLK_SIZE;
+    if (fill_file_with_repeated_bytes(fd, len, 0xFF) < 0) {
+        THROW_ERROR("failed to fill file");
+    }
+    int mode = FALLOC_FL_KEEP_SIZE | 0xDEAD;
+    off_t offset = 0;
+    len = len / 2;
+    int ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EOPNOTSUPP) {
+        THROW_ERROR("failed to check fallocate with invalid mode");
+    }
+    mode = FALLOC_FL_PUNCH_HOLE | FALLOC_FL_ZERO_RANGE;
+    ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EOPNOTSUPP) {
+        THROW_ERROR("failed to check fallocate with punch_hole and zero_range");
+    }
+    mode = FALLOC_FL_PUNCH_HOLE;
+    ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EOPNOTSUPP) {
+        THROW_ERROR("failed to check punch_hole without keep_size");
+    }
+    mode = FALLOC_FL_INSERT_RANGE | FALLOC_FL_KEEP_SIZE;
+    ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EINVAL ) {
+        THROW_ERROR("failed to check insert range with other flags");
+    }
+    mode = FALLOC_FL_COLLAPSE_RANGE | FALLOC_FL_KEEP_SIZE;
+    ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EINVAL ) {
+        THROW_ERROR("failed to check collapse range with other flags");
+    }
+    mode = FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE | FALLOC_FL_UNSHARE_RANGE;
+    ret = fallocate(fd, mode, offset, len);
+    if (ret >= 0 || errno != EINVAL ) {
+        THROW_ERROR("failed to check unshare flag without allocate mode");
+    }
```

Comment:
I think writing a loop that iterates an array of input arguments and expected output is more concise and readable. And also easier to add new test cases in the future.

Reply:
Good suggestion